### PR TITLE
configurable number prefix without mouse grid

### PR DESCRIPTION
**Pull request by jaresty**
- Attempt to implement optional numb prefix
- The tags need to be logically combined
- Respond to suggestion that the mouse grid should not be considered

### CODE REVIEW

**Comment by fidgetingbits** [comment]
Thanks for your patience. I've pushed some changes that I think are more in line with what @auscompgeek originally
suggested in #943, and that will be easier to maintain. I'll detail the changes a bit below, and why I did it this way
instead of how you did it, in case you're interested:

- The name of the 'numb' prefix is no longer in the tag name. It's possible people would change the prefix to something else, so should never include the actual spoken form in the tag name imo. The name "disable_unprefixed_numbers" also fits nicely in that mouse_grid setting the tag becomes more self-explanatory.
- Set the tag in mouse_grid, so that it still works
- Move the unprefixed version of the command in it's own file, rather than the prefixed version. This is because many users add other types of numbers (hexadecimal, etc) to numbers.talon already, which also have prefixes of various forms. So just isolating the non-prefixed version seems cleaner.
- Add documentation to the settings file so that users now how/why to use the tag.

**Comment by fidgetingbits** [comment]
I guess one remaining question is if we should set the tag by default or not.

File: settings.talon
Comment (auscompgeek) [inline settings.talon:76]: Shouldn't this example be a `tag()` rather than setting a setting?

**Comment by jaresty** [comment]
my preference would be for the recommended behavior (I.e. not requiring the prefix) be the default.

File: settings.talon
Comment (fidgetingbits) [reply settings.talon]: Thanks

**Review by pokey — COMMENTED** [review]
this looks ok to me. tempted to have it on by default, but then we'd need to do some kind of deprecation process

also, double negatives always confuse me. I wonder if we should have a `allow_unprefixed_numbers` tag that is turned on by default in settings, and the user can comment it out if they don't want unprefixed numbers

**Comment by fidgetingbits** [comment]
> also, double negatives always confuse me. I wonder if we should have a `allow_unprefixed_numbers` tag that is turned on by default in settings, and the user can comment it out if they don't want unprefixed numbers

I considered that but since mouse grid needs to disable non prefixed numbers to use its own nonprefixed number commands, it seemed simpler to have a disable tag. 

Maybe there is a good way for mouse grid to remove the tag when it activates? I don't recall off hand if you can just shut off a a tag like that.

**Comment by fidgetingbits** [comment]
fwiw I'm happy to have prefixed the default as well. 

random thought, but for deprecating stuff I almost wonder if we should introduce a deprecated notification action that pops a notification only the first time you use a deprecated command after starting talon (or every N times maybe), but that also spams the logs on every use. that way it's easier for less technical users to find out (since I suspect many don't read the logs), but still in a way thats not too spammy. it would be useful here and for future depreciation I suspect.

**Comment by pokey** [comment]
> for deprecating stuff I almost wonder if we should introduce a deprecated notification action that pops a notification only the first time you use a deprecated command

I believe that kind of thing is already covered by our [deprecation infrastructure / policy decoumentation](https://github.com/talonhub/community/blob/main/core/deprecations.py)

File: core/numbers/numbers.py
Comment (knausj85) [inline core/numbers/numbers.py:180]: I might suggest just calling this "prefixed_numbers"

**Comment by knausj85** [comment]
I agree the prefix should be the default to improve compatibility with cursorless  / reduce misrecognitions

File: core/numbers/numbers_unprefixed.talon
Comment (knausj85) [inline core/numbers/numbers_unprefixed.talon:1]: I believe we still want
 and not tag: user.mouse_grid_showing
here

**Comment by pokey** [comment]
> I agree the prefix should be the default to improve compatibility with cursorless / reduce misrecognitions

makes sense, tho in that case we should prob use command deprecation infra so it doesn't just break when user updates?

File: core/numbers/numbers_unprefixed.talon
Comment (fidgetingbits) [reply core/numbers/numbers_unprefixed.talon:1]: I don't think so because mouse_grid is updated to set the tag, which would disable numbers_unprefixed.talon. As per #943 the idea of doing it the way I did was that the core/numbers code doesn't need to be aware of the existence of mouse grid which I think is preferable. This way mouse grid itself can just toggle the functionality off, so its own functionality won't conflict. 

The downside of this is that it's a double negative tag. To me having core/numbers not having to special case for some plugin is worth the slightly confusing double negative tag, but if others disagree then can change it.

File: core/numbers/numbers.py
Comment (fidgetingbits) [reply core/numbers/numbers.py:180]: See other comment about why it's this way.

**Comment by fidgetingbits** [comment]
> > for deprecating stuff I almost wonder if we should introduce a deprecated notification action that pops a notification only the first time you use a deprecated command
> 
> I believe that kind of thing is already covered by our [deprecation infrastructure / policy decoumentation](https://github.com/talonhub/community/blob/main/core/deprecations.py)

Thanks for the reminder. I just added the relevant updates. Of course wording and such will be subject what tag ends up being settled on.

File: core/numbers/numbers.py
Comment (auscompgeek) [reply core/numbers/numbers.py:180]: Could call it "numbers must be prefixed" if the double negative is too confusing?

Comment (fidgetingbits) [reply core/numbers/numbers.py:180]: Ok, I'll update it to be non-double negative tag then. user.prefix_numbers or user.numbers_must_be_prefixed? I think I prefer the latter because it's a bit more self explanatory, but good with whatever others prefer.

File: core/numbers/numbers_unprefixed.talon
Comment (pokey) [inline core/numbers/numbers_unprefixed.talon:5]:
```suggestion
    user.deprecate_command("2024-01-27", "<number>", "numb <number>")
```

File: settings.talon
Comment (pokey) [inline settings.talon:78]:
```suggestion
# "numb one" to write "1". Note that this tag will eventually be activated by default
```

**Review by pokey — COMMENTED** [review]
Looks good! Left a couple final minor comments

**Comment by jaresty** [comment]
Anything left to fix on this? Wanted to make sure you weren't waiting on me 😀

**Comment by fidgetingbits** [comment]
I was mostly waiting for someone else to actually approve after I addressed the last of the suggested changes, but I suppose since pokey said it was fine with the minor changes I can approve it myself.

**Review by fidgetingbits — APPROVED** [review]

**Pull request merged**
